javascript/ql/src/Security/CWE-830/FunctionalityFromUntrustedSource.ql

```diff
@@ -0,0 +1,57 @@
+/**
+ * @name Inclusion of untrusted functionality by a HTML element.
+ * @description Including untrusted functionality by a HTML element
+ *              opens up for potential man-in-the-middle attacks.
+ * @kind problem
+ * @problem.severity warning
+ * @security-severity 8.1
+ * @precision high
+ * @id js/functionality-from-untrusted-source
+ * @tags security
+ *       external/cwe/cwe-830
+ */
+
+import javascript
+import semmle.javascript.HTML
+
+bindingset[host]
+predicate isAllowedHost(string host) { host.toLowerCase().regexpMatch("localhost(:[0-9]+)?/.*") }
+
+bindingset[path]
+predicate isUntrustedSourcePath(string path) {
+  path.substring(0, 2) = "//"
+  or
+  exists(string hostPath | hostPath = path.regexpCapture("http://(.*)", 1) |
+    not isAllowedHost(hostPath)
+  )
+}
+
+abstract class IncludesUntrustedContent extends HTML::Element {
+  IncludesUntrustedContent() { this = this }
+
+  /** Gets an explanation why this source is untrusted. */
+  abstract string getProblem();
+}
+
+/** A script element that refers to untrusted content. */
+class ScriptElementWithUntrustedContent extends IncludesUntrustedContent, HTML::ScriptElement {
+  ScriptElementWithUntrustedContent() {
+    isUntrustedSourcePath(this.getSourcePath()) and
+    not exists(string digest | not digest = "" | this.getIntegrityDigest() = digest)
+  }
+
+  override string getProblem() {
+    result = "script elements should use an https link and/or use the integrity attribute"
+  }
+}
+
+/** An iframe element that includes untrusted content. */
+class IframeElementWithUntrustedContent extends HTML::IframeElement, IncludesUntrustedContent {
+  IframeElementWithUntrustedContent() { isUntrustedSourcePath(this.getSourcePath()) }
+
+  override string getProblem() { result = "iframe elements should use an https link" }
+}
+
+from IncludesUntrustedContent s, string problem
```

Comment:
Food for thought: can we easily extend the query with dynamically generated elements with unsafe sources?

```
e.src = "http://example.com/cdn/script.js"
```

Reply:
Wow, I built a(n additional) taint tracking query and it's finding quite a lot of things:
https://lgtm.com/query/6174245023330040398/

Should I be adding something like that to this PR, or make a fresh one? There's a risk the PR might become unwieldy — but it's up to the reviewers :)

Reply:
That branch could, of course also be grown further, like by adding support for jquery..

Reply:
If the addition is as simple as the linked query suggests, I think we should get in now to ensure the various phrasings a general enough from the start.

Could we perhaps start the addition in a more precise mode than proposed on LGTM? The pattern `('https:' == document.location.protocol ? 'https://' : 'http://') + '<host>.com/js/<script>.js'` seems common in the results, but these days it is probably always ending up in the `https`-branch since the user likely is visiting the web page on HTTPS (and if he isn't other MITM attacks are possible anyway!).

Reply:
I think it makes sense to flag the ternary — just like we also flag URLs starting with `//` (protocol-relative URLs).

I've added the new query to this PR.

Reply:
Isn't the user already vulnerable to MITM if the current page is served as `http`? Tbh I don't think we should flag these cases.

Reply:
Technically, yes. but an attacker is more likely to target, eg., jquery than a web page using jquery (jquery is a high value target). Or, similarly: a user using the page via http, but the page fetching via https is vulnerable in one way; if the page is also using http, the user is vulnerable in two ways.

The reason the `//` strings are flagged is because Erik asked me to add that behaviour yesterday — I'm now stuck with conflicting input, @asgerf @esbena.

Reply:
OK, I'll remove that behaviour.